text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,1417 @@
+
+A `BatchArrowRecords` message is composed of 3 attributes. The protobuf definition is:
+
+```protobuf
+message BatchArrowRecords {
+  // [mandatory] Batch ID. Must be unique in the context of the stream.
+  string batch_id = 1;
+
+  // [mandatory] A collection of payloads containing the data of the batch.
+  repeated OtlpArrowPayload otlp_arrow_payloads = 2;
+}
+```
+
+The `batch_id` attribute is a unique identifier for the batch inside the scope of the current stream. It is used to
+uniquely
+identify the batch in the egress `BatchStatus` stream. See the [Batch Id generation](#batch-id-generation) section for
+more information on the implementation of this identifier.
+
+The `otlp_arrow_payloads` attribute is a list of `OtlpArrowPayload` messages. Each `OtlpArrowPayload` message represents
+a table of data encoded in a columnar format (e.g. metrics, logs, traces, and future OTLP entities). Several correlated
+IPC Arrow messages of different nature and with different schemas can be sent in the same OTLP batch identified by
+`batch_id` and thus be processed as one unit without complex logic in the collector or any other processing systems.
+More details on the `OtlpArrowPayload` columns in the section [Mapping OTEL entities to Arrow records](#mapping-otel-entities-to-arrow-records).
+
+More specifically, an `OtlpArrowPayload` protobuf message is defined as:
+
+```protobuf
+message OtlpArrowPayload {
+  // [mandatory] A unique id assigned to a sub-stream of the batch sharing the same schema, and dictionaries.
+  string sub_stream_id = 1;
+
+  // [mandatory] Type of the OTLP Arrow payload.
+  OtlpArrowPayloadType type = 2;
+
+  // [mandatory] Serialized Arrow Record Batch
+  // For a description of the Arrow IPC format see: https://arrow.apache.org/docs/format/Columnar.html#serialization-and-interprocess-communication-ipc
+  bytes record = 4;
+
+  // [mandatory]
+  CompressionMethod compression = 5;
+}
+
+enum OtlpArrowPayloadType {
+  METRICS = 0;
+  LOGS = 1;
+  SPANS = 2;
+}
+
+message EncodedData {
+  bytes ipc_message = 1;
+  bytes arrow_data = 2;
+}
+
+enum CompressionMethod {
+  NO_COMPRESSION = 0;
+  ZSTD = 1;
+}
+```
+
+The `sub_stream_id` attribute is a unique identifier for a sub-stream of `BatchArrowRecords` sharing the same schema inside
+the
+scope of the current stream. This id will be used receiver side to keep track of the schema and dictionaries for a
+specific
+type of Arrow Records. See the [Substream Id generation](#substream-id-generation) section for more information on the
+implementation
+of this identifier.
+
+The `OtlpArrowPayloadType` enum specifies the `type` of the payload.
+
+The `record` attribute is a binary representation of the Arrow RecordBatch.
+
+The `compression` attribute is a mandatory attribute that is used to define the compression algorithms for the different
+bytes buffer.
+
+ZSTD offers a training mode, which can be used to tune the algorithm for a selected type of data. The result of this
```

Comment:
Do you suggest that a training is performed at build time or during the protocol design, and resulting dictionary is hard-coded in this protocol's implementations?

Reply:
No, I'm simply describing how ZSTD is designed by default, and how we could potentially improve the compression ratio for small batches in the future. We could achieve this by using ZSTD's ability to train its internal dictionary, thereby optimizing compression for these specific situations.